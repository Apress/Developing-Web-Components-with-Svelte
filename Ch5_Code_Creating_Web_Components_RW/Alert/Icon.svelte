<svelte:options tag="cobalt-icon" />

<script>
  export let width = "24px";
  export let height = "24px";
  export let iconType = "";

  let icons = [
    {
      box: 512,
      name: "info",
      svg: `<path d="M248 64C146.39 64 64 146.39 64 248s82.39 184 184 184 184-82.39 184-184S349.61 64 248 64z" fill="none" stroke="currentColor" stroke-miterlimit="10" stroke-width="32"/><path fill="none" stroke="#ffffff" stroke-linecap="round" stroke-linejoin="round" stroke-width="32" d="M220 220h32v116"/>
      <path fill="none" stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="32" d="M208 340h88" /><path d="M248 130a26 26 0 1026 26 26 26 0 00-26-26z" fill="#ffffff" />`,
    },
    {
      box: 512,
      name: "warn",
      svg: `<path d="M448 256c0-106-86-192-192-192S64 150 64 256s86 192 192 192 192-86 192-192z" stroke="#000000" fill="none" stroke-miterlimit="10" stroke-width="32"/><path d="M250.26 166.05L256 288l5.73-121.95a5.74 5.74 0 00-5.79-6h0a5.74 5.74 0 00-5.68 6z" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="32"/><path d="M256 367.91a20 20 0 1120-20 20 20 0 01-20 20z" fill="#000000" />`,
    },
  ];

  let displayIcon = icons.find((e) => e.name === iconType);
</script>

<svg
  class={$$props.class}
  {width}
  {height}
  viewBox="0 0 {displayIcon.box} {displayIcon.box}"
>
  {@html displayIcon.svg}
</svg>
