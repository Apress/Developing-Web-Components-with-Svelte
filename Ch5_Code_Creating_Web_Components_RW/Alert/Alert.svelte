<svelte:options tag="cobalt-alert" />

<script>
  import Icon from "./Icon.svelte";

  export let show;
  export let icon;
  export let close;
  export let type = "";
  export let title = "";
  export let description = "";
  let showAnimation = true;
  let typeClass;

  // Convert string value to boolean where appropriate
  let showIcon = JSON.parse(icon);

  switch (type) {
    case "warn":
      typeClass = "alert-warn";
      break;
    case "dark":
      typeClass = "alert-dark";
      break;
    case "error":
      typeClass = "alert-error";
      break;
    case "info":
      typeClass = "alert-info";
      break;
    case "success":
      typeClass = "alert-success";
      break;
    default:
      typeClass = "";
  }
  const classes = ["alert", typeClass, showAnimation ? "fade-in" : ""]
    .filter((klass) => klass.length)
    .join(" ");

  const closeAlert = () => {
    show = false;
  };
</script>

{#if show}
  <dialog open on:click={close} class={classes} role="alert">
    <div class="icon">
      {#if showIcon}<Icon iconType={type} />{/if}
    </div>
    <div class="message">
      <strong>
        {title}
      </strong>
      {description}
    </div>
    <div>
      <button on:click={closeAlert}>&#x2716;</button>
    </div>
  </dialog>
{/if}

<style>
  dialog {
    min-width: 300px;
    display: flex;
    justify-content: space-between;
    font-family: Arial, Helvetica, sans-serif;
    border: none;
  }

  button {
    background: none;
    border: none;
    font-size: 21px;
  }

  .icon {
    margin-right: 10px;
  }

  .message {
    display: flex;
    flex-direction: column;
    line-height: 24px;
    min-width: 300px;
  }

  .fade-in {
    animation: fade-in 2000ms both;
  }

  @keyframes fade-in {
    from {
      opacity: 0%;
    }
  }

  .alert-warn {
    background: #ffeb3b;
    color: #000000;
  }

  .alert-info {
    background: #2196f3;
    color: #ffffff;
  }
</style>
