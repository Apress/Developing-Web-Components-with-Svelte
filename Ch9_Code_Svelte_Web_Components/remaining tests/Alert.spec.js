import { render, screen } from '@testing-library/svelte';
import { fireEvent } from '@testing-library/dom';
import Alert from "../src/lib/Alert/Alert.svelte";

jest.useFakeTimers();

describe("Tests for Alert", () => {
  const props = {
		show: true,
		description: "An info description",
		title: "Simple Info",
		icon: "true",
		type: "info",
    close: "false",
  };

  it("should render properly", async() => {
    const result = await(render(Alert, { props }));
    expect(() => result).not.toThrow();
  });

  // https://testing-library.com/docs/svelte-testing-library/example
  it("should disappear when close button clicked", async() => {
    const result = await(render(Alert, { props }));
    const alertDialog = fireEvent.click(screen.getByRole("alert"));
    expect(alertDialog).tobeNull;
  })

  it("get a snapshot of component", () => {
    const tree = render(Alert, props);
    expect(tree).toMatchSnapshot();
  })
});