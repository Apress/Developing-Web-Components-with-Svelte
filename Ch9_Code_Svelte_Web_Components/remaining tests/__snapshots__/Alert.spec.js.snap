// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Alert get a snapshot of component 1`] = `
{
  "component": Alert {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {},
      "ctx": [
        true,
        "false",
        "info",
        "Simple Info",
        "An info description",
        true,
        "alert alert-info fade-in",
        [Function],
        "true",
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {
        "close": 1,
        "description": 4,
        "icon": 8,
        "show": 0,
        "title": 3,
        "type": 2,
      },
      "root": <div>
        <dialog
          class="alert alert-info fade-in svelte-p9d6qo"
          open=""
          role="alert"
        >
          <div
            class="icon svelte-p9d6qo"
          >
            <svg
              height="24px"
              viewBox="0 0 512 512"
              width="24px"
            >
              <path
                d="M248 64C146.39 64 64 146.39 64 248s82.39 184 184 184 184-82.39 184-184S349.61 64 248 64z"
                fill="none"
                stroke="currentColor"
                stroke-miterlimit="10"
                stroke-width="32"
              />
              <path
                d="M220 220h32v116"
                fill="none"
                stroke="#ffffff"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
              
      
              <path
                d="M208 340h88"
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-miterlimit="10"
                stroke-width="32"
              />
              <path
                d="M248 130a26 26 0 1026 26 26 26 0 00-26-26z"
                fill="#ffffff"
              />
            </svg>
          </div>
           
          <div
            class="message svelte-p9d6qo"
          >
            <strong>
              Simple Info
            </strong>
             
            An info description
          </div>
           
          <div>
            <button
              class="svelte-p9d6qo"
            >
              ✖
            </button>
          </div>
        </dialog>
        
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <dialog
        class="alert alert-info fade-in svelte-p9d6qo"
        open=""
        role="alert"
      >
        <div
          class="icon svelte-p9d6qo"
        >
          <svg
            height="24px"
            viewBox="0 0 512 512"
            width="24px"
          >
            <path
              d="M248 64C146.39 64 64 146.39 64 248s82.39 184 184 184 184-82.39 184-184S349.61 64 248 64z"
              fill="none"
              stroke="currentColor"
              stroke-miterlimit="10"
              stroke-width="32"
            />
            <path
              d="M220 220h32v116"
              fill="none"
              stroke="#ffffff"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="32"
            />
            
      
            <path
              d="M208 340h88"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-miterlimit="10"
              stroke-width="32"
            />
            <path
              d="M248 130a26 26 0 1026 26 26 26 0 00-26-26z"
              fill="#ffffff"
            />
          </svg>
        </div>
         
        <div
          class="message svelte-p9d6qo"
        >
          <strong>
            Simple Info
          </strong>
           
          An info description
        </div>
         
        <div>
          <button
            class="svelte-p9d6qo"
          >
            ✖
          </button>
        </div>
      </dialog>
      
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
